Introduce virtual provider uuid

michaelkuhn opened the pull request:
libuuid is currently contained in util-linux, libuuid and uuid. This change introduces a new virtual provider `uuid` and renames the existing `uuid` package to `ossp-uuid`.

util-linux's libuuid is provided in the form of a separate package util-linux-uuid to make sure that packages depending on uuid and util-linux can use a separate uuid implementation, which the concretizer does not allow if libuuid is contained in util-linux.

Fixes #10812

michaelkuhn committed:
Introduce virtual provider uuid

libuuid is currently contained in util-linux, libuuid and uuid. This
change introduces a new virtual provider `uuid` and renames the existing
`uuid` package to `ossp-uuid`.

util-linux's libuuid is provided in the form of a separate package
util-linux-uuid to make sure that packages depending on uuid and
util-linux can use a separate uuid implementation, which the concretizer
does not allow if libuuid is contained in util-linux.

michaelkuhn commented:
cc @lee218llnl @carns @adamjstewart

adamjstewart commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -162,7 +162,7 @@ class Python(AutotoolsPackage):
         # the easy solution is to only depend on Spack's libuuid when *not* on
         # a Mac.
-        depends_on('libuuid', when='+uuid')
+        depends_on('uuid', when='+uuid')
```
Can you confirm that `python+uuid` builds correctly with a few different `uuid` providers on Linux? I'm planning on making `python+uuid` the default in #18252.

michaelkuhn replied:
I did a test build but of course forgot to actually specify `+uuid`. :man_facepalming: I just pushed a new version that should fix this.

adamjstewart commented:
> util-linux's libuuid is provided in the form of a separate package util-linux-uuid to make sure that packages depending on uuid and util-linux can use a separate uuid implementation, which the concretizer does not allow if libuuid is contained in util-linux.

Is this a use case that we really care about?

michaelkuhn commented:
> Is this a use case that we really care about?

I am not sure, to be honest. There are a few other benefits that I forgot to mention, though: It allows us to get rid of util-linux's python dependency, which would result in a circular dependency. The only way around that was to disable Python for +libuuid or to have a "build only libuuid" variant, which also felt wrong.

adamjstewart commented on var/spack/repos/builtin/packages/bcache/package.py:
```diff
@@ -19,7 +19,7 @@ class Bcache(MakefilePackage):
 
-    depends_on('libuuid')
+    depends_on('uuid')
     depends_on('util-linux')
```
This was updated in a separate PR. It had 2 deps that provided libuuid, and now only has one. Should still be converted to a virtual provider.

michaelkuhn replied:
I fixed the conflict and re-added the `uuid` dependency, since libuuid is not provided by util-linux anymore.

carns commented:
Thanks for working on this @michaelkuhn .  I haven't been able to test it yet but I plan to do so soon.

carns commented:
FYI, when I build with a `uuid` dependency in this configuration I get the following during the build of `util-linux-uuid`:

```
  >> 645    /usr/bin/install: cannot remove '/usr/share/bash-completion/complet
            ions/uuidgen': Permission denied
```

My test system happens to have a native `util-linux` package installed already.  It looks like the make install target for the spack build tried to update some things in the system path, though from surrounding output it looks like most things went in the spack package path as expected.

michaelkuhn commented:
Strange, this does not happen for me. Do you see the same error for the util-linux package?

carns commented:
Good call to check that; I actually get a long list of errors similar to that one when just installing util-linux:

```
  >> 1100    /usr/bin/install: cannot remove '/usr/share/bash-completion/comple
             tions/blkid': Permission denied
  >> 1101    /usr/bin/install: cannot create regular file '/usr/share/bash-comp
             letion/completions/cal': Permission denied
  >> 1102    /usr/bin/install: cannot create regular file '/usr/share/bash-comp
             letion/completions/col': Permission denied
  >> 1103    /usr/bin/install: cannot create regular file '/usr/share/bash-comp
             letion/completions/colcrt': Permission denied
...
```
I guess that's a separate github issue then, doesn't have much to do with your PR :)

carns commented:
Well, this has been reported before and fixed apparently: https://github.com/spack/spack/issues/7632

adamjstewart commented:
@michaelkuhn I would like to push this PR forward again. Lately, I've been encountering a lot of issues where libuuid is listed as a transitive dependency and ends up breaking my builds on macOS. The cause is already listed in our `python` package:

> On macOS systems, Spack's libuuid conflicts with the system-installed version and breaks anything linked against Cocoa/Carbon. Since the system-provided version is sufficient to build Python's UUID support, the easy solution is to only depend on Spack's libuuid when *not* on a Mac.

There are two solutions:

1. do what we do in `python` and don't include a libuuid dependency on macOS for every package that depends on libuuid
2. create a `apple-uuid` package for the system package and add is as an external package by default

1 is quick and easy but annoying and tedious. I would rather go with 2. We already do this for `apple-libunwind`. Of course, 2 requires this PR first.

michaelkuhn commented:
I have rebased this PR to current develop since there were a few conflicts. I have also added uuid dependencies to all packages that depend on util-linux (which previously included libuuid), so we do not break anything.

michaelkuhn commented:
I did a few test builds and everything seems to work fine. I noticed an oddity with python, though: Switching the uuid provider does not seem to make a difference and even if I comment out the libuuid dependency, I get an importable uuid module (even without these changes). :confused:

adamjstewart commented:
Python will pick up the system uuid libs, especially on macOS.

adamjstewart commented on etc/spack/defaults/packages.yaml:
```diff
@@ -49,6 +49,7 @@ packages:
       unwind: [libunwind]
       yacc: [bison, byacc]
       flame: [libflame, amdlibflame]
+      uuid: [util-linux-uuid, libuuid]
```
It looks like most packages previously depended on `libuuid`. Any reason why you chose `util-linux-uuid` as the default? The fact that it includes `linux` in the name concerns me as a macOS user.

michaelkuhn replied:
Mostly due to https://github.com/spack/spack/issues/10812#issuecomment-590625762, which states that the util-linux version is the only one still maintained. I don't have any strong preference, though.

adamjstewart replied:
Sounds like a good enough reason to me!

adamjstewart commented:
Do we want to create an `apple-libuuid` package now or in a future PR? I'm not actually sure where the library lives on macOS or if it is always present, or if it is the same as any of these other providers.

michaelkuhn commented:
> Python will pick up the system uuid libs, especially on macOS.

My OS Python has native uuid libraries (`/usr/lib64/python3.9/lib-dynload/_uuid.cpython-39-x86_64-linux-gnu.so`) and importing the uuid module also makes Python dynamically load the system's libuuid (checked via `/proc/id/maps`). This does not happen with the Spack Python but the uuid module still seems to work. :man_shrugging:

michaelkuhn commented:
> Do we want to create an `apple-libuuid` package now or in a future PR? I'm not actually sure where the library lives on macOS or if it is always present, or if it is the same as any of these other providers.

A separate PR might be easier since I don't have access to macOS and someone else would have to take care of such a package.

adamjstewart commented on var/spack/repos/builtin/packages/util-linux-uuid/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    version('2.36', sha256='82942cd877a989f6d12d4ce2c757fb67ec53d8c5cd9af0537141ec5f84a2eea3')
+
+    depends_on('pkgconfig')
```
```suggestion
    depends_on('pkgconfig', type='build')
```

michaelkuhn replied:
Fixed

adamjstewart commented on var/spack/repos/builtin/packages/util-linux/package.py:
```diff
@@ -27,9 +28,6 @@ class UtilLinux(AutotoolsPackage):
     depends_on('python@2.7:')
     depends_on('pkgconfig')
```
```suggestion
    depends_on('python@2.7:', type='build')
    depends_on('pkgconfig', type='build')
```

michaelkuhn replied:
Fixed
